distro/limits.go

```diff
@@ -0,0 +1,151 @@
+		setDefault: func(limits trace.SpanLimits) trace.SpanLimits {
+			// Unlimited.
+			limits.EventCountLimit = -1
+			return limits
+		},
+	},
+
+	// Maximum allowed attributes per span event.
+	{
+		envs: []string{spanEventAttributeCountKey},
+		setDefault: func(limits trace.SpanLimits) trace.SpanLimits {
+			// Unlimited.
+			limits.AttributePerEventCountLimit = -1
+			return limits
+		},
+	},
+
+	// Maximum allowed span link count.
+	{
+		envs: []string{spanLinkCountKey},
+		setDefault: func(limits trace.SpanLimits) trace.SpanLimits {
+			limits.LinkCountLimit = 1000
+			return limits
+		},
+	},
+
+	// Maximum allowed attributes per span link.
+	{
+		envs: []string{spanLinkAttributeCountKey},
+		setDefault: func(limits trace.SpanLimits) trace.SpanLimits {
+			// Unlimited.
+			limits.AttributePerLinkCountLimit = -1
+			return limits
+		},
+	},
+}
+
+// spanLimitSources is a collection of all the sources of configuration for
+// a group of spans limits.
+type spanLimitSources []spanLimitSource
+
+// applyDefaults will apply the defaults to all limits that do not have the
+// associated limit's environment variables set.
+func (s spanLimitSources) applyDefaults(limits trace.SpanLimits) trace.SpanLimits {
+	for _, source := range s {
+		limits = source.applyDefault(limits)
+	}
+	return limits
+}
+
+// spanLimitSource are all the sources a span limit is configured from and the
+// default value used if none of those sources are used.
+type spanLimitSource struct {
+	envs       []string
+	setDefault func(limits trace.SpanLimits) trace.SpanLimits
+}
+
+// envsSet returns if any environment variable for s is set.
+func (s spanLimitSource) envsSet() bool {
+	for _, env := range s.envs {
+		if _, ok := os.LookupEnv(env); ok {
+			return true
+		}
+	}
+	return false
+}
+
+// applyDefaults will apply the default to the corresponding limit in limits
+// if the associated environment variables are not set.
+func (s spanLimitSource) applyDefault(limits trace.SpanLimits) trace.SpanLimits {
+	if !s.envsSet() {
+		// Panic if setDefault is nil. It never should be, and if it is alert
+		// the developer that made the change as soon as possible.
+		limits = s.setDefault(limits)
+	}
+	return limits
+}
+
+// newSpanLimits returns new span limits that use Splunk defaults (the link
+// count is limited to 1000, the attribute value length is limited to 12000,
+// and all other limts are set to be unlimited) or the corresponding OTel
+// environment variable value if it is set.
+func newSpanLimits() *trace.SpanLimits {
+	// Use trace.NewSpanLimits here to ensure any future additions are not set
+	// to zero, which would happen if we delared with &trace.SpanLimits{...}.
+	limits := trace.NewSpanLimits()
+
+	// limits will use OTel defaults or the applicable environment variable if
+	// they are set. The Splunk defaults need to be applied only if the
+	// environment variables were unset.
+	limits = sources.applyDefaults(limits)
+
+	return &limits
+}
```

Comment:
It took me a while to understand this code and to be certain that is OK 😉 

I propose to simplify it to something like this:

```golang
func newSpanLimits() *trace.SpanLimits {
	// Use trace.NewSpanLimits here to ensure any future additions are not set
	// to zero, which would happen if we delared with &trace.SpanLimits{...}.
	limits := trace.NewSpanLimits()

	applyDefault(&limits.AttributeValueLengthLimit, 12000, attributeValueLengthKey, spanAttributeValueLengthKey)
	applyDefault(&limits.AttributeCountLimit, -1, attributeCountKey, spanAttributeCountKey)
	applyDefault(&limits.EventCountLimit, -1, spanEventCountKey)
	applyDefault(&limits.AttributePerEventCountLimit, -1, spanEventAttributeCountKey)
	applyDefault(&limits.LinkCountLimit, 1000, spanLinkCountKey)
	applyDefault(&limits.AttributePerLinkCountLimit, -1, spanLinkAttributeCountKey)

	return &limits
}

// applyDefault sets val to dft if none of envVars  is defined.
func applyDefault(val *int, dft int, envVars ...string){
	for _, env := range envVars {
		if _, ok := os.LookupEnv(env); ok {
			return
		}
	}
	*val = dft
}
```

Reply:
I prefer the current implementation. It encapsulates related values with the type system instead of parameters to a function call, it has methods related to a type defined for that type, uses the language to loop instead of writing each iteration in code, declares the associated limit sources as a package level variable instead of containing them in a function, and doesn't use functions to modify a passed argument instead of using a return value.

That said, I care more about progressing this than keeping with the current approach. I can update if it will achieve this goal, with the caveat that `applyDefaults` will be updated to return a value instead of modifying a passed argument.

Reply:
Updated.

Reply:
> declares the associated limit sources as a package level variable

Side note: I dislike global variables and I avoid them as much as possible 😉 I will probably always complain when I see one - it is not personal 😄   

> I care more about progressing

👍 

> with the caveat that applyDefaults will be updated to return a value instead of modifying a passed argument.

That is a good one.